finality-aleph/src/network/rmc.rs

```diff
@@ -1,6 +1,99 @@
+//! A set of abstractions for dealing with `ReliableMulticast` in a more testable
+//! and modular way.
+//!
+//! We expose the following traits:
+//! - `Multicast`: mimicking the interface of `aleph_bft::ReliableMulticast`
+//! - `Multisigned`: abstracting away the `aleph_bft::Multisigned` capabilities as a trait.
+//!
+//! For use in production code, we provide wrappers for the BFT code pieces:
+//! - `RMCWrapper` that simply dispatches all calls to its internal `ReliableMulticast` instance
+//! - `MultisignedWrapper` that does the same for `Multisigned`.
+
 use crate::{aggregator::SignableHash, crypto::Signature};
-use aleph_bft::{rmc::Message, SignatureSet};
+use aleph_bft::rmc::ReliableMulticast;
+use aleph_bft::{
+    rmc::Message, MultiKeychain, Multisigned as MultisignedStruct, SignatureSet, UncheckedSigned,
+};
+use codec::Codec;
 use sp_runtime::traits::Block;
+use std::fmt::Debug;
+
+/// A convenience trait for gathering all of the desired hash characteristics.
+pub trait Hash: AsRef<[u8]> + std::hash::Hash + Eq + Clone + Codec + Debug + Send + Sync {}
+
+impl<T: AsRef<[u8]> + std::hash::Hash + Eq + Clone + Codec + Debug + Send + Sync> Hash for T {}
 
 pub type NetworkData<B> =
     Message<SignableHash<<B as Block>::Hash>, Signature, SignatureSet<Signature>>;
+
+/// Anything that exposes the same interface as `aleph_bft::Multisigned` struct.
+/// In production code this will be the `MultisignedWrapper`, containing an instance
+/// of `aleph_bft::Multisigned` (here imported as `MultisignedStruct` for clarity).
+pub trait Multisigned<'a, H: Hash, MK: MultiKeychain> {
+    fn as_signable(&self) -> &SignableHash<H>;
+    fn into_unchecked(self) -> UncheckedSigned<SignableHash<H>, MK::PartialMultisignature>;
+}
+
+pub struct MultisignedWrapper<'a, H: Hash, MK: MultiKeychain> {
```

Comment:
I don't think you need the wrapper (and the other one either), is there a reason why you cannot implement this trait directly for `MultisignedStruct`?

Reply:
Two reasons: 1) I was told it was the correct way of doing it :joy:  2) I wasn't sure how Rust would be able to dispatch to the right method. Removed the wrappers and it works fine, thanks! :)